uplc-semantics.md

```diff
@@ -409,6 +412,70 @@ module UPLC-SEMANTICS
 	    #then (True)
 	    #else (False)
 	    #fi) ... </k>
+
+  // sha3_256
+  // UPLC bytestrings are almost what K's Sha3_256 needs. However, a proper string needs to
+  // be built from the ground up. This is done with the following steps:
+  // 1. We convert the given ByteString to a string representing an hex number and then do its
+  //    decimal representation.
+  //    `String2Base(trimByteString(B), 16)`.
+  // 2. The integer from step 1. is converted into a Byte representation,
+  //    with possible leading zeros preserved.
+  //    `Int2Bytes(lengthString(trimByteString(B)) /Int 2, StringBase(...), BE)`,
+  //    where the expression `lengthString(trimByteString(B)) /Int 2` preserves the leading zeros.
+  // 3. The Bytes resulting from step 2 are then translated into a string that can be consumed
+  //    by `Sha3_256`.
+  // 4. The last step simply converts the string resulting from step 3 into a ByteString.
+  
+  rule <k> (builtin sha3_256 .TermList) =>
+           (con bytestring String2ByteString("#" +String
+                Sha3_256(Bytes2String(Int2Bytes(lengthString(trimByteString(B)) /Int 2,
+                                      String2Base(trimByteString(B), 16 ), BE))))) ... </k>
+       <stack> ... (ListItem((con bytestring B:ByteString)) => .List) </stack>
+
+  rule <k> (builtin sha3_256) => #SHA3 ... </k>
+
+  rule <k> (V:Value ~> ([ Clos(#SHA3, _RHO) _])) => #SHA3(V) ... </k>
+
+  rule <k> #SHA3((con bytestring B:ByteString)) =>
+           (con bytestring String2ByteString("#" +String
+                Sha3_256(Bytes2String(Int2Bytes(lengthString(trimByteString(B)) /Int 2,
+                                      String2Base(trimByteString(B), 16), BE))))) ... </k>
```

Comment:
There's some code duplication between the RHS of this rule and the RHS of the rule at line 430. I wonder if you could create a function to reduce the duplication..

Reply:
I think you could add a rule like this:
```
rule formatByteStringForSha(x) => Bytes2String(Int2Bytes(lengthString(trimByteString(x)) /Int 2,
                                      String2Base(trimByteString(x), 16), BE))) [macro]
```
and use this macro formatByetStringForSha on the rule RHS and then write the documentation for steps 1-4 above the macro definition. You might want to use a different name for the rule if you find a more suitable name for it.

if you do this, then the rule can look like this:

```
rule <k> #SHA3((con bytestring B:ByteString)) =>
         (con bytestring String2ByteString("#" +String Sha3_256(formatByteStringForSha(B)))
```

Maybe you might consider writing a helper function called `String2#ByteString` so that you can hide the "#" with a macro like this
```
rule String2#ByteString(x) => String2ByteString("#" +String x) [macro]
```
and use it like so:
```
rule <k> #SHA3((con bytestring B:ByteString)) =>
        (con bytestring String2#ByteString(Sha3_256(formatByteStringForSha(B)))
```
This last macro part might not be necessary but in general, it would be easier to read code with a little less duplication.

Reply:
All builtins are like this. There could be some more general structure that could simplify the code but they are not just too many and I think there is no reason to think about general coding patterns for something so simple. Moreover, the semantics will change when we start working with proofs. The simpler the code the easier to prove.

Reply:
> All builtins are like this. There could be some more general structure that could simplify the code but they are not just too many and I think there is no reason to think about general coding patterns for something so simple. Moreover, the semantics will change when we start working with proofs. The simpler the code the easier to prove.

Maybe this interface isn't the best way to communicate this. I'll try to explain with a little more context. I might repeat some of the same things but I want to make sure that we're both on the same page. After reviewing this pull request, I found 6 instances of duplication (with each builtin having 2 duplicates). It looks like what you intended was for all 6 rules to use the same thing:
```
Bytes2String(Int2Bytes(lengthString(trimByteString(x)) /Int 2, String2Base(trimByteString(x), 16), BE)))
```
I think duplicating this 6 times is a lot and when you have to change the semantics, you have to change it in 6 places. I'm not sure how proving would work. Could you explain to me why leaving this as-is would make proofs easier as opposed to refactoring common parts to a different rule? We can chat internally as well if you think that's easier